install phpactor with phpactor-update function

kermorgant opened the pull request:
The recent evolutions in phpactor's rpc protocol made me realize that non BC changes could also break the correction operation of this package (ie `version` and `force_reload` new parameters).

Thus, I believe the simplest and safest approach to install phpactor *for its use from phpactor.el* is to have it installed from phpactor.el (with its version controlled  in composer.json).

I would also qualify the use of an externally installed phpactor as risky, and therefore require the user to configure ``phpactor-executable`` manually in order to draw attention on this.

solves #59

kermorgant commented on phpactor.el:
```diff
@@ -84,11 +84,21 @@
   (or (when phpactor-executable
         (php-project--eval-bootstrap-scripts phpactor-executable))
-      (executable-find phpactor-command-name)
-      (let ((vendor-executable (f-join (phpactor-get-working-dir) "vendor/bin/phpactor")))
```
I was a bit puzzled by this way of installing phactor (inside a php project), but even so, I'd rather recommend everyone to either install it from phpactor.el or define it manually (hopefully after having considered the risks it brings).

how does that sound ?

kermorgant commented:
Maybe we should clone phpactor's git repo before in order to provide a more accurate information to phpactor-status

![image 3](https://user-images.githubusercontent.com/10561580/45600226-da991800-ba01-11e8-8185-83e48f181b65.png)

kermorgant commented:
renamed phpactor-install to phpactor-update. That way, it mimics the vim plugin and in the long run, the name will make more sense as user will have to re-run regularly.

kermorgant commented:
I'd like to have a way to run `phpactor-update` upon after installation or upgrade of this package, but I've not found anything in that regard. 

But I suspect that if something existed, it would be something related to some package manager and that we should limit ourself to some configuration suggestion.

zonuexe commented on phpactor.el:
```diff
@@ -86,11 +86,21 @@
+      (setq default-directory package-folder)
+      (call-process "composer" nil (get-buffer-create phpactor-action--buffer-name) nil "install" "--no-dev")))
+
+(defun phpactor-package-folder ()
```
Use `directory` instead of `folder`.

zonuexe commented on phpactor.el:
```diff
@@ -460,7 +470,9 @@ function."
-    (error "Phpactor uses rpc protocol %s whereas this package requires %s" version phpactor--supported-rpc-version))
+    (if (phpactor-executable)
+        (error "Phpactor uses rpc protocol %s but this package requires %s" version phpactor--supported-rpc-version)
+        (error "Phpactor should be upgraded. Please run phpactor-update")))
```
These lines can be modified as follows, but what do you think?

```diff
@@ -469,10 +469,11 @@ function."
 ;; Dispatcher:
 (cl-defun phpactor-action-dispatch (&key action parameters version)
   "Execite action by `NAME' and `PARAMETERS'."
-  (when (and version (not (equal phpactor--supported-rpc-version version)))
-    (if (phpactor-executable)
+  (unless (or (null version)
+              (version<= phpactor--supported-rpc-version version))
+    (if (phpactor-find-executable)
         (error "Phpactor uses rpc protocol %s but this package requires %s" version phpactor--supported-rpc-version)
-        (error "Phpactor should be upgraded. Please run phpactor-update")))
+      (error "Phpactor should be upgraded.  Please run phpactor-update")))
   (phpactor--add-history 'phpactor-action-dispatch (list :action action :parameters parameters :version version))
   (let ((func (cdr-safe (assq (intern action) phpactor-action-table))))
     (if func
```

kermorgant replied:
well, I'm not yet used to `unless` so it takes me a little bit more effort to read...

>version<= phpactor--supported-rpc-version version

Not sure I see what you mean here. if versions are equal, no issue but if `version` is different than `phpactor--supported-rpc-version version`, I think we should stop with an error

zonuexe replied:
Thanks, I see.

zonuexe commented on phpactor.el:
```diff
@@ -86,11 +86,21 @@
         (when (file-exists-p vendor-executable)
           vendor-executable))))
 
+(defun phpactor-update ()
```
How about naming `phpactor-setup` instead of `phpactor-update`?

kermorgant replied:
hmm, I tend to like the semantic that comes with the term `update` as it conveys the idea that sooner or later, you'll have to run it again to stay up to date. 

But I agree that for the first time you use it, it feels a bit wrong. Would that justify having both name, one aliasing towards the other ?

zonuexe commented:
@kermorgant I think that this change is almost no problem. 👍 

Perhaps it may be helpful for you, I just made this change to `composer.el`: https://github.com/emacs-php/composer.el/pull/4/files#diff-b6acbc3fae6805c6382033710ab67382R226

kermorgant commented:
@zonuexe thanks for the review. Indeed, ensuring composer is installed via composer.el seems good !

kermorgant commented on phpactor.el:
```diff
@@ -86,11 +87,22 @@
+  (interactive)
+  (let ((package-folder (phpactor-package-directory)))
+      (setq default-directory package-folder)
+      (composer-ensure-exist-managed-composer-phar)
```
@zonuexe I thought this would work but composer.phar was not installed in the expected folder (~.emacs.d/var)

kermorgant replied:
just had a look at it and seems like installing composer programmatically on arch linux can be problematic (opened this [issue](https://github.com/emacs-php/composer.el/issues/5))

=> This needs a bit different approach and a some error checking

kermorgant replied:
I did not solve my issue about installing composer programmatically but I opted to use existing composer first and try to install it if it is not found.

@zonuexe I believe I answered all your comments. However, I'm not sure if dependency towards composer.el is properly defined

ejmr commented:
Am I understanding correctly that the plan moving forward is for `composer.el` and `phpactor.el` to themselves install the appropriate versions of Composer and Phpactor?

kermorgant commented:
@ejmr 
> Am I understanding correctly [...] ?

Yes. 

I currently have some concerns about composer.el which sould be installed and enabled transparently as a dependency but that's not the case with my setup (maybe because I'm working with the sources of phpactor.el). Without that, I doubt composer.el sould be used.


Regarding the installation of phpactor, I noticed that even minor changes in the rpc protocol (such as the addition of a 'version' information) would break some of phpactor.el's operations. 

Having phpactor.el installing phpactor avoids such issues.  Installation is also simpler that way imo.

ejmr commented:
That all sounds reasonable and understandable.  I feel like we should consider offering users a way to opt-out and use a global installation of Composer and/or Phpactor if possible, if only because I know some people will be certain to complain about their installation despite all good intentions.

kermorgant commented:
@zonuexe I changed small bits (fix for default directory during composer invocation) and added an explicit opt-out variable for composer installation.  Do you see something still needing rework ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
Having thought a bit more about the installation of composer through composer.el, I find it a bit questionable as phpactor's ideal use-case is for composer based projects.

So I'd assume a phpactor user to have composer installed globally. Therefrom, I'd just make the ``phpactor-update`` function fail if composer executable is not found in $PATH. 

How does that sound ?

kermorgant commented:
@zonuexe no worry, thx
